### PR TITLE
Join oracle data to specs with a join table

### LINKED ISSUE #9294: [Bug]: Datanode panic when using a lot of ethereum-oracles
### Problem encountered

Data node panic:
```
2023-09-05T11:03:04.972Z	INFO	datanode.networkHistory.snapshot	snapshot/service_create_snapshot.go:36	creating snapshot for {Network History Segment for Chain ID:testnet-001 Height From:7789 Height To:7810}
2023-09-05T11:03:04.973Z	INFO	datanode.networkHistory.snapshot	snapshot/service_create_snapshot.go:115	copying all table data....
2023-09-05T11:03:05.555Z	ERROR	datanode.sqlstore-broker	broker/sqlstore_broker.go:350	failed to process event id:"7812-439"  block:"10D5E247E3FC832CCDB12AB32A5D683FB4F7F21FC9969F342BCC19F6EFFE6103"  type:BUS_EVENT_TYPE_ORACLE_DATA  oracle_data:{external_data:{data:{data:{name:"vegaprotocol.builtin.timestamp"  value:"1693911784"}  data:{name:"vegaprotocol.builtin.timetrigger"  value:"1693911784"}  matched_spec_ids:"3174ef78e15899e166b7d82dbdf5bbf428232d5c3b1bb8bcfa0282572ec3c39b"  matched_spec_ids:"26c3364e8e87fc3c64aec93fdfa13fc0ed9824ecd7d098826c6360add496cb3f"  matched_spec_ids:"4ed6ff76c020e4cb9ffd6c5ca515c5034907133c34ca9daa724fc6c402e40ec5"  matched_spec_ids:"511c488ddb8831c48997561112f85c52da64b2b2f6e18b6945cbb70332d83479"  matched_spec_ids:"2d0d9a2c974fe9ca29d5b57132f1541776f8b4519e6fbc1e4a4abc17abdcdfd0"  matched_spec_ids:"364c8704da4fc65393098f6a1584f556729468da81707735ea0ddf20662e5f72"  matched_spec_ids:"bfa4a4a4a45e2859daf76066b90d56a6c399cd06dec6ad0445922e1009620cfd"  matched_spec_ids:"f2e5c384bc46dd3ad4656331e1c5b95f4e6d08c7bc5c75d1b7653c298b34eb1f"  matched_spec_ids:"bce5cfd82d8972b5e5ad87d25559aea30f22e8dd265359a959028ca9496085bb"  matched_spec_ids:"c3d671c7aea99ef69710578b64956d3227375fa8a997b41f6a2ed2a96d0c4d35"  matched_spec_ids:"5be588575583431ac97cb7afe9d8ae63ad26a5e4ce2dfcf029e8c5b7d20815bf"  matched_spec_ids:"08ca0aa38ee3a362bda29967ce18c9ac2c47081505d26c7982da3ade3482b016"  matched_spec_ids:"301d9f5d46f01e33b832e10664c816d9f5cab509e1ed17d1f19f0cdf9e4d1daa"  matched_spec_ids:"0ea81f84a1ee87511ba0d23fdcee530309d66bdef7c06c0addd2da21e9737c39"  matched_spec_ids:"646590f515cdff8d1973a2d090ba0b41e75aa20f75b7a56b4aacbc01ced750f7"  matched_spec_ids:"e859dcf159fbd3ba2c6ae2e71bfb3fed41219dfb9251b85cb5d10510d1db2a1e"  matched_spec_ids:"140c18b99f9daf4ae4b7f146f355defd45498a68e57d0479dc1a4378d006627b"  matched_spec_ids:"399716dde8d02d337ae3dcbf70cb648e93330cf0522b7d02e1578ad1cbe902c6"  matched_spec_ids:"f660efa63b29d2088639e6cda5db47aeff95482102f9a63cf7266ad76fbedcbc"  matched_spec_ids:"e0ce8316e21f5eddd154edaf50ceebe060f8d35b9bc0bfbe35c3762400e3264f"  matched_spec_ids:"0f53ed9a17993c113f62f02b255d0a66bf14630a66166c6546848f13be662e09"  matched_spec_ids:"41b1d8964e4fe9ef9c27b656c6e0c1739a5b12332028ef23a7cdf193d0c2d29d"  matched_spec_ids:"0b43f9b7886acb3fe65d6b80c3f284dea33b46b956e556190bc673a1189ff1e9"  matched_spec_ids:"525456338fbe0acb8278feaca865708cbbc9f322cd2f859c000f8ebdb0bc542f"  matched_spec_ids:"39eb3874916fc56929bc8664602dbfad7cdd3981085bacbebd6ffa4e519987f1"  matched_spec_ids:"39d00f782d006d4e88999fb427e375d9c1c379ca0db6e23bc1594e717a42499e"  matched_spec_ids:"16377a4253908dc49da487246bc635f496dee448087499635d677c3802671aff"  matched_spec_ids:"68e36a7af139709c7856e2a444cd6748e594c6caa61e24152625c8792f9a24e8"  matched_spec_ids:"e34913ac227b12df232b35012a85d0e850b1138b5914abc0590d84d61fe47619"  matched_spec_ids:"7c7eafc302b427c635ee33e60c0b3c515fb7767b802e3a996984a5496f6a589c"  matched_spec_ids:"67e1fb9fda87ea4c0840cbcea1c14484d97d0a2406f7f871aa58fc557554f26a"  matched_spec_ids:"9fcd44ed413f027b1527e275e6dd331d96e620efd7f4c130b12d8319aa0f9d86"  matched_spec_ids:"b72d49d4f535debd1a61b4f1b6ecc140dba4c53e01e42ed6f87d841a387c202d"  matched_spec_ids:"9afd816ae86f936fc3628a38c7d20e71263c940722985e84e16f3670747ab421"  matched_spec_ids:"0b149cbd8e7521695aca78ba6a27a269b4fd34172bbb39f1ab77ebe34a637e1d"  matched_spec_ids:"b5e7dac195220e84a5227e72cb176aa572fd3ab5f1e69d96fa52cb0fcffe0206"  matched_spec_ids:"ae1ce8f795d848414053b9ffb4fd6f8f7f06663757491667ad2f684bbfafa1e0"  matched_spec_ids:"134fa1b17e2a5b35635d56ecf2619cef6fe39e02ab4543403e189b21354ca147"  matched_spec_ids:"0b5afa94e4ab128d06a5156ae1ca4c21a4ddc95c0565d8c0f99d55dfe2d8b5d8"  matched_spec_ids:"bcf02fc2b1bc5510f7a14e7ba5e887090b5724ad98605383b143e918b52ec1a6"  matched_spec_ids:"676ac31fa3ebeaa755011a79fe02836504cca3028c900c757700e8c44458b07b"  matched_spec_ids:"2da14fca0d2e4ee68decd0d982fcf2b3057ffd3df2fe6147870018a4fb870e19"  matched_spec_ids:"cef59db20697bc7bc0463c106d05bcd0e6bdc1bdc173b585da04b2b8637e98cf"  matched_spec_ids:"60a18a7cb0a9185fb031066499010487b70db3466bce502b32b828bdef9a1c4a"  matched_spec_ids:"7a6b2920c9aec81ca7a7faaa9d12348c808aaa9cb8d124403e510569ab71bf63"  matched_spec_ids:"a749aa5cba5e28b8530ac93b5a786a6493554ec3fbb625fa23ee7543cc113432"  matched_spec_ids:"73e4d2d4205600af9cf645fe1bfd5b104fdea43d5c055980fccab79a8dcbbced"  matched_spec_ids:"a5406b7f4d4cf8f6a51f64abed46e6349fc7ae063428f2a74a0bd097985364d2"  matched_spec_ids:"f5eb75397021737bd31073e96c686c23439cc79c3e39248858fb6e317337038e"  matched_spec_ids:"dfd18da1bceed0f3596e4dd4028b763372d7a2115fe4a36e2962461f4142a3ea"  matched_spec_ids:"29763513069841710e2b6ba69c9caf297d03f6a4c8ba70e3b9c5e704f68fe50f"  matched_spec_ids:"70a9ac7a6b42d2f651782abbd06ec610059ef806e148628229d14787e374c2c8"  matched_spec_ids:"95fbc30201c6325b9edf22f110115feee19837e52c05954fd5d5b8bf69dfadb0"  matched_spec_ids:"1618b1825e31fd6c80ddcfc39df040159a4ad62be154ebc44ba3ad37231e989b"  matched_spec_ids:"c52cf766e38450d7df49f71a80b414b68cb04703520f936534fb845bcf7ad766"  matched_spec_ids:"e8ee6076306f2afba9050bc922d6ec08cee1e0d68874a97278ce794a2b263434"  matched_spec_ids:"559328ef8bc3feaf29b747f67398d440a8a0abe7965ab2e09cd653f373733370"  matched_spec_ids:"d17b186a2a05197cf60ad941d8d2e58e0099d489ce6db792bad81728b3c5b0af"  matched_spec_ids:"e493dd3406d49aa62ea5984fb2336e3954636c8f8250414d390b8612842063e4"  matched_spec_ids:"66a40d1b16f8c5db8d1a8718120fb19647edf88dbeda36a5c37b6a7302e628fa"  matched_spec_ids:"e9b45e73f647fa60a1c0b9a2d6019faab3de32f4b71578e53cae217b7862dd52"  matched_spec_ids:"8ee07f6d1f7c0c889de07c68e4ccfe641e49b6c4d064542b281027daf6b4f442"  matched_spec_ids:"dd6bc9ff06bfda5e8ac152031b5875ad92937fcf2e2d56c06984e4d9017aa6af"  matched_spec_ids:"81d20406d77f6510dc129286969bc535dc77d70256c9846b121bf05181039dea"  matched_spec_ids:"98f0c49a768b9d3be68de4408159b5e702e09ac991d929ba091cd79992390885"  matched_spec_ids:"78c1fedaa0d3315965729a869466a16394a37b3143ec886ab7423c4dd198fbb7"  matched_spec_ids:"033f6f570d1b523550e2c8453a397747053a06840c95d7cc460eb68c5b61eb13"  matched_spec_ids:"4a3f71204e49fadb4ebf4c2671cad295e203f5fdab64ab73c17c14ccba938371"  matched_spec_ids:"3c02fa9b85b8d4e31c03cdaa1f29b5d4b1e1ce4c4e3c5531fccb007ccd1c136a"  matched_spec_ids:"0ed6129dc7f1b1479c09df9138aafe385a6093d2b9bbaf958f5d1ecced8bf475"  matched_spec_ids:"e113c7d9ad2011bb0e6cfa04401af544d737d2cc1468c899eee83e58479ef4d6"  matched_spec_ids:"5acc16bbc6707dc4307f8a3c504448846e5765ae09a64c3392f4cb050745e1ce"  matched_spec_ids:"e9340348c8148fcb84010e905e3f231fe17890a43b69cc859111074d392aba2f"  matched_spec_ids:"1a563483d736f1cad7079c4e31c3e0a84d56177a36fe2f58f596c7e0627a07ce"  matched_spec_ids:"35d778055b6d5cdf3bbd90a7b92d6e99a3df5cd133a0c39193bdac8b689494c8"  matched_spec_ids:"8011b620fd660da02208d0bc009751880203fedb44ee423f3bb932e5d7c1b7d6"  matched_spec_ids:"82fe5b7cb47ed7bb224793593c952d0dd48e3e09399e6931b049a7557e23767c"  matched_spec_ids:"50bfb584b356a51263518ff8e832dff38ab706f0d0ace741717b1e7ffabe7275"  matched_spec_ids:"0ae2c6b13ff7fec5f44cb4b0b87f691084402a2bad98a14da5a1e70e6524402d"  broadcast_at:1693911784748337403}}}  version:1  chain_id:"testnet-001"  tx_hash:"10D5E247E3FC832CCDB12AB32A5D683FB4F7F21FC9969F342BCC19F6EFFE6103" error:could not insert oracle data into database: ERROR: index row size 2880 exceeds btree version 4 maximum 2704 for index "idx_oracle_data_matched_spec_ids" (SQLSTATE 54000)
```

which occurred when I kicked off a full system test run with all market set to perpetuals, so we make lots of them. I'm guessing its because they are all using the same contract to call into we're getting into this situation where we have a lot of matched spec ids.

It would be good to understand the limitation here.

Note that this does not block perpetual testing because we can:
- use a signer key in all default perpetual markets when doing a full-run-as-perps
- make sure we terminate all perpetual markets in perpetual tests that do need eth-oracles

Jenkins run is here:
https://jenkins.vega.rocks/job/common/job/system-tests-wrapper/3583/

### Observed behaviour

Data node panic

### Expected behaviour

Not data node panic?

### Steps to reproduce

```bash
kick off a full system test run with all markets as perpetuals
```

### Software version

develop

### Failing test

_No response_

### Jenkins run

https://jenkins.vega.rocks/job/common/job/system-tests-wrapper/3583/

### Configuration used

_No response_

### Relevant log output

_No response_

### PR DESCRIPTION
**Pull request by pscott31**
also upgrade oracle_data to be a hypertable to prevent snapshot bloat

closes #9294

### CODE REVIEW

**Review by jeremyletang — CHANGES_REQUESTED** [review]
New migration have been added in the meantime, could you rename this one 0033_* (or so if others have been added since this message), thanks.

File: datanode/sqlstore/migrations/0034_oracle_data.sql [hunk 1]
```diff
@@ -0,0 +1,69 @@
+-- +goose Up
+-- First make oracle_data a hypertable
+-- +goose StatementBegin
+DO $$
+BEGIN
+    IF NOT EXISTS (SELECT * FROM timescaledb_information.hypertables WHERE hypertable_name = 'stop_orders') THEN
```
Comment (guoguojin) [inline datanode/sqlstore/migrations/0034_oracle_data.sql:6]: You're checking for the stop_orders hypertable here, I'm not sure this is correct.

Comment (pscott31) [reply datanode/sqlstore/migrations/0034_oracle_data.sql:6]: tbh I wasn't sure what that was for but figured you had had some fun experience to lead you to put it there..

File: protos/sources/vega/checkpoint/v1/checkpoint.proto [hunk 2]
```diff
@@ -213,69 +213,87 @@ message MarketTracker {
...
+  repeated ReturnsData returns_data = 12;
 
-  DataWithHistory total_maker_fees_received = 16;
-  DataWithHistory total_maker_fees_paid = 17;
-  DataWithHistory total_lp_fees = 18;
+  repeated EpochPartyFees maker_fees_received_history = 13;
+  repeated EpochPartyFees maker_fees_paid_history = 14;
+  repeated EpochPartyFees lp_fees_history = 15;
+
+  repeated EpochTimeWeightPositionData time_weighted_position_data_history = 16;
```
Comment (candida-d) [inline protos/sources/vega/checkpoint/v1/checkpoint.proto:232]: Should this be EpochTimeWeight**ed**PositionData?

**Review by guoguojin — DISMISSED** [review]

File: protos/sources/vega/checkpoint/v1/checkpoint.proto (hunk 2 above)
Comment (pscott31) [reply protos/sources/vega/checkpoint/v1/checkpoint.proto]: That huge list of changes appeared when I messed up my rebase; that change is actually already committed

File: datanode/sqlstore/migrations/0036_oracle_data.sql [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+        PRIMARY KEY (vega_time, seq_num, spec_id)
+    );
+
+CREATE INDEX IF NOT EXISTS idx_oracle_data_oracle_specs_spec_to_data ON oracle_data_oracle_specs (spec_id);
+
+-- Make it a hypertable
+SELECT
+    create_hypertable ('oracle_data_oracle_specs', 'vega_time', chunk_time_interval => INTERVAL '1 day');
+
+-- Populate it with the contents of the 'matched_specd_ids' bytea array column
```
Comment (candida-d) [inline datanode/sqlstore/migrations/0036_oracle_data.sql:27]:
```suggestion
-- Populate it with the contents of the 'matched_specs_ids' byte array column
```

Comment (pscott31) [reply datanode/sqlstore/migrations/0036_oracle_data.sql:27]: That one is fair enough though it should actually be matched_spec_ids

Comment (pscott31) [reply datanode/sqlstore/migrations/0036_oracle_data.sql]: fixed

Comment (candida-d) [reply datanode/sqlstore/migrations/0036_oracle_data.sql]: wasn't sure which way it should be, thanks for fixing

Comment (candida-d) [reply datanode/sqlstore/migrations/0036_oracle_data.sql]: @pscott31 "bytea" is irrelevant I guess? Or is that how it should be?

Comment (pscott31) [reply datanode/sqlstore/migrations/0036_oracle_data.sql]: yes that's the name of the database type; odd as it is

**Review by candida-d — DISMISSED** [review]

**Review by guoguojin — APPROVED** [review]

**Review by jeremyletang — APPROVED** [review]

**Pull request merged**